Realm/RLMSyncSessionRefreshHandle.mm

```diff
@@ -33,40 +33,132 @@ @interface RLMSyncSessionRefreshHandle () {
 @property (nonatomic) NSTimer *timer;
 
 @end
 
 @implementation RLMSyncSessionRefreshHandle
 
-- (instancetype)initWithFullURLPath:(NSString *)urlPath
+- (instancetype)initWithPathToRealm:(NSString *)path
                                user:(RLMSyncUser *)user
                             session:(std::shared_ptr<realm::SyncSession>)session {
     if (self = [super init]) {
-        self.fullURLPath = urlPath;
+        self.pathToRealm = path;
         self.user = user;
-        _session = session;
+        _session = std::move(session);
         return self;
     }
     return nil;
 }
 
-- (void)invalidate {
+- (void)dealloc {
     [self.timer invalidate];
-    self.user = nil;
 }
 
-- (void)scheduleRefreshTimer:(NSTimeInterval)fireTime {
-    static const NSInteger refreshBuffer = 10;
+- (void)invalidate {
     [self.timer invalidate];
-    self.timer = [[NSTimer alloc] initWithFireDate:[NSDate dateWithTimeIntervalSince1970:(fireTime - refreshBuffer)]
-                                          interval:0
-                                            target:self
-                                          selector:@selector(timerFired:)
-                                          userInfo:nil
-                                           repeats:NO];
-    [[NSRunLoop currentRunLoop] addTimer:self.timer forMode:NSDefaultRunLoopMode];
+}
+
++ (NSDate *)fireDateForTokenExpirationDate:(NSDate *)date {
+    static NSTimeInterval refreshBuffer = 10;
+    NSDate *fireDate = [date dateByAddingTimeInterval:-refreshBuffer];
+    // Only fire times in the future are valid.
+    return ([fireDate compare:[NSDate date]] != NSOrderedDescending ? fireDate : nil);
+}
+
+- (void)scheduleRefreshTimer:(NSDate *)dateWhenTokenExpires {
+    // Schedule the timer on the main queue.
+    // It's very likely that this method will be run on a side thread, for example
+    // on the thread that runs `NSURLSession`'s completion blocks. We can't be
+    // guaranteed that there's an existing runloop on those threads, and we don't want
+    // to create and start a new one if one doesn't already exist.
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [self.timer invalidate];
+        // The fire time is `self.refreshBuffer` seconds before the token expires.
+        NSDate *fireDate = [RLMSyncSessionRefreshHandle fireDateForTokenExpirationDate:dateWhenTokenExpires];
+        if (!fireDate) {
+            [self.user _unregisterRefreshHandleForURLPath:self.pathToRealm];
+            return;
+        }
+        NSTimer *t = [[NSTimer alloc] initWithFireDate:fireDate
+                                              interval:0
+                                                target:self
+                                              selector:@selector(timerFired:)
+                                              userInfo:nil
+                                               repeats:NO];
+        [[NSRunLoop currentRunLoop] addTimer:t forMode:NSDefaultRunLoopMode];
```

Comment:
Unless `_unregisterRefreshHandleForURLPath` needs to be run on the main thread, this can skip the dispatch to the main thread and just do `[[NSRunLoop mainRunLoop] addTimer:...`